src/can.rs

```diff
@@ -0,0 +1,111 @@
+//! Controller Area Network
+
+use core::convert::TryInto;
+use core::convert::TryFrom;
+
+use nb;
+
+/// A type that can either be `BaseId` or `ExtendedId`
+#[cfg(feature = "unproven")]
+pub trait Id: Sized {
+    /// The BaseId variant
+    type BaseId: BaseId + Into<Self> + TryFrom<Self>;
+
+    /// The ExtendedId variant
+    type ExtendedId: ExtendedId + Into<Self> + TryFrom<Self>;
+}
+
+/// A Can (11-bit) ID
+#[cfg(feature = "unproven")]
+pub trait BaseId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Extended (28-bit) ID
+#[cfg(feature = "unproven")]
+pub trait ExtendedId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Frame
+#[cfg(feature = "unproven")]
+pub trait Frame {
+    /// The Id type of this Frame
+    type Id: Id;
```

Comment:
I'm dubious of the value of having Frame type and the Id type be associated.  It's very common for both Id types to be present on the bus at the same time.

Reply:
This is not a problem as Id is defined as "A type that can either be `BaseId` or `ExtendedId`"

Reply:
Just one Frame type and one ID type. Some examples of how the Frame type could look (only one of them, is actually required):
``` rust
pub enum Frame1 {
    BaseIdDataFrame(_),
    BaseIdRemoteFrame(_),
    ExtendedIdDataFrame(_),
    ExtendedIdRemoteFrame(_),
}

pub struct Frame2 {
    remote: bool,
    extended_id: bool,
    data: [u8; 8],
}
```

and the Id type could look like (only one of them, is actually required):

``` rust
extern crate ux;
use ux::{u11, u29};

pub enum Id1 {
    Base(u11),
    Extended(u29),
}

pub struct Id2 {
    extended: bool,
    id: u32,
}
```

When implementing the trait each representation of a `Frame` would have to choose a representation of a n`Id`.

Reply:
So you would also have 

    impl BaseId for Id2 {...}
    impl ExtendedId for Id2 {...}

?